### PR TITLE
improve abstract detection

### PR DESCRIPTION
**Pull request by jens1o**

### CODE REVIEW

**Comment by jens1o** [comment]
I decided to not use the `getText()` function, because they are only 2 possibilities and it would slow down the process...

File: src/DefinitionResolver.php [hunk 1]
```diff
@@ -1190,6 +1194,11 @@ private function tryGetDocBlockTagForParameter($docBlock, $variableName)
             if ($tag->getVariableName() === \ltrim($variableName, "$")) {
                 return $tag;
             }
+        } else if ($node instanceof Node\Expr\FuncCall && $node->name instanceof Node\Name && strtolower((string)$node->name) === 'define') {
+            if (!isset($node->args[0]) || !($node->args[0]->value instanceof Node\Scalar\String_)) {
+                return null;
+            }
+            return (string)$node->args[0]->value->value;
```
Comment (felixfbecker) [inline src/DefinitionResolver.php]: bad merge?

File: src/DefinitionResolver.php [hunk 2]
```diff
@@ -174,7 +174,11 @@ public function createDefinitionFromNode(Node $node, string $fqn = null): Defini
         $def->fqn = $fqn;
 
         // Determines whether the suggestion will show after "new"
-        $def->canBeInstantiated = $node instanceof Node\Statement\ClassDeclaration;
+        $def->canBeInstantiated = (
+            $node instanceof Node\Statement\ClassDeclaration &&
+            // use text's length for speedup: |abstract| = 8; |final| = 5
+            ($node->abstractOrFinalModifier === null || $node->abstractOrFinalModifier->length !== 8)
```
Comment (felixfbecker) [inline src/DefinitionResolver.php]: that doesn't look faster than comparing it to `TokenKind::AbstractKeyword`

Comment (jens1o) [inline src/DefinitionResolver.php]: Good catch, I didn't thought about this.

File: src/DefinitionResolver.php [hunk 3]
```diff
@@ -174,7 +174,11 @@ public function createDefinitionFromNode(Node $node, string $fqn = null): Defini
         $def->fqn = $fqn;
 
         // Determines whether the suggestion will show after "new"
-        $def->canBeInstantiated = $node instanceof Node\Statement\ClassDeclaration;
+        $def->canBeInstantiated = (
+            $node instanceof Node\Statement\ClassDeclaration &&
+            // check whether it is not an abstract class
+            ($node->abstractOrFinalModifier === null || $node->abstractOrFinalModifier->kind !== PhpParser\TokenKind::AbstractKeyword)
```
Comment (felixfbecker) [inline src/DefinitionResolver.php]: You can remove the check for `null`

Comment (jens1o) [inline src/DefinitionResolver.php]: No, I can't. Because when there is no modifier, it returns null.

Comment (jens1o) [inline src/DefinitionResolver.php]: image

Comment (felixfbecker) [inline src/DefinitionResolver.php]: In which case `$node->abstractOrFinalModifier->kind !== PhpParser\TokenKind::AbstractKeyword` is `true`

Comment (jens1o) [inline src/DefinitionResolver.php]: That will result in a `access to undefined property 'kind'`?

Comment (felixfbecker) [inline src/DefinitionResolver.php]: sorry, I thought `$kind` was `null`.

**Pull request merged**